File: 1472-design-browser-history/1472-design-browser-history.cpp
class Node {
public:
    string val;
    Node* next;
    Node* prev;
    
    Node(string val){
        this->val = val;
        this->next = NULL;
        this->prev = NULL;
    }
};

class BrowserHistory {
public:
    Node* cur;
    BrowserHistory(string homepage) {
        cur = new Node(homepage);
    }
    
    void visit(string url) {
        Node* nn = new Node(url);
        cur->next = nn;
        nn->prev = cur;
        cur = cur->next;
    }
    
    string back(int steps) {
        
        while(cur->prev && steps > 0){
            cur = cur->prev;
            steps -= 1;
        }
        
        return cur->val;
    }
    
    string forward(int steps) {
        while(cur->next && steps > 0){
            cur = cur->next;
            steps -= 1;
        }
        
        return cur->val;
    }
};

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory* obj = new BrowserHistory(homepage);
 * obj->visit(url);
 * string param_2 = obj->back(steps);
 * string param_3 = obj->forward(steps);
 */

File: 1472-design-browser-history/1472-design-browser-history.java
class Node {
    
    public String value;
    public Node next;
    public Node prev;
    
    Node(String value){
        this.value = value;
        this.next = null;
        this.prev = null;
    }
}

class BrowserHistory {
    private Node cur; 
    
    public BrowserHistory(String homepage) {
        cur = new Node(homepage);
    }
    
    public void visit(String url) {
        Node newPage = new Node(url);
        cur.next = newPage;
        newPage.prev = cur;
        cur = cur.next;
    }
    
    public String back(int steps) {
        
        while(cur.prev != null && steps > 0){
            cur = cur.prev;
            steps -= 1;
        }
        
        return cur.value;
    }
    
    public String forward(int steps) {
        while(cur.next != null && steps > 0){
            cur = cur.next;
            steps -= 1;
        }
        
        return cur.value;
    }
}

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory obj = new BrowserHistory(homepage);
 * obj.visit(url);
 * String param_2 = obj.back(steps);
 * String param_3 = obj.forward(steps);
 */